File: leetcode91/leetcode91_1.cpp
class Solution {
public:
    int numDecodings(string s) {
        int n = s.length();
        
        vector<int> dp(n + 1, 0); // dp[i] means the ways from s[i] to the end of s
        // dp[n] is empty '', which is one way
        if (n == 0)
            return 0;
        
        dp[n] = 1;
        if (s[n - 1] != '0')
            dp[n - 1] = 1;
        
        for (int i = n - 2; i >= 0; i--) {
            if (s[i] == '0')
                dp[i] = 0;
            else {
                dp[i] = dp[i + 1];
                
                if (s[i] == '1' || (s[i] == '2' && s[i + 1] <= '6'))
                    dp[i] += dp[i + 2];
            }
        }
        
        return dp[0];
    }
};

// Runtime: 4 ms, faster than 65.13% of C++ online submissions for Decode Ways.
// Memory Usage: 8.4 MB, less than 88.24% of C++ online submissions for Decode Ways.
// time complexity: O(n)
// space complexity: O(n)

File: leetcode91/leetcode91_1.java
class Solution {

    Map<Integer, Integer> memo = new HashMap<>();

    public int numDecodings(String s) {
        return recursiveWithMemo(0, s);
    }

    private int recursiveWithMemo(int index, String str) {
        // Have we already seen this substring?
        if (memo.containsKey(index)) {
            return memo.get(index);
        }

        // If you reach the end of the string
        // Return 1 for success.
        if (index == str.length()) {
            return 1;
        }

        // If the string starts with a zero, it can't be decoded
        if (str.charAt(index) == '0') {
            return 0;
        }

        if (index == str.length() - 1) {
            return 1;
        }

        int ans = recursiveWithMemo(index + 1, str);
        if (Integer.parseInt(str.substring(index, index + 2)) <= 26) {
            ans += recursiveWithMemo(index + 2, str);
        }

        // Save for memoization
        memo.put(index, ans);

        return ans;
    }
}

// Solution 1
// Runtime: 2 ms, faster than 35.02% of Java online submissions for Decode Ways.
// Memory Usage: 37.8 MB, less than 33.06% of Java online submissions for Decode
// Ways.
// time complexity: O(n)
// space complexity: O(n)